trio/_subprocess_platform/waitid.py

```diff
@@ -0,0 +1,116 @@
+    # is bigger so that's what we use to size `pad`; while
+    # there are a few extra fields in there, most of it is
+    # true padding which would not be written by the syscall.
+    waitid_ffi.cdef(
+        """
+typedef struct siginfo_s {
+    int si_signo;
+    int si_errno;
+    int si_code;
+    int si_pid;
+    int si_uid;
+    int si_status;
+    int pad[26];
+} siginfo_t;
+int waitid(int idtype, int id, siginfo_t* result, int options);
+"""
+    )
+    waitid = waitid_ffi.dlopen(None).waitid
+
+    def sync_wait_reapable(pid):
+        P_PID = 1
+        WEXITED = 0x00000004
+        if sys.platform == 'darwin':  # pragma: no cover
+            # waitid() is not exposed on Python on Darwin but does
+            # work through CFFI; note that we typically won't get
+            # here since Darwin also defines kqueue
+            WNOWAIT = 0x00000020
+        else:
+            WNOWAIT = 0x01000000
+        result = waitid_ffi.new("siginfo_t *")
+        while waitid(P_PID, pid, result, WEXITED | WNOWAIT) < 0:
+            got_errno = waitid_ffi.errno
+            if got_errno == errno.EINTR:
+                continue
+            raise OSError(got_errno, os.strerror(got_errno))
+
+
+# adapted from
+# https://github.com/python-trio/trio/issues/4#issuecomment-398967572
+
+waitid_limiter = CapacityLimiter(math.inf)
+
+
+async def _waitid_system_task(pid: int, event: Event) -> None:
+    """Spawn a thread that waits for ``pid`` to exit, then wake any tasks
+    that were waiting on it.
+    """
+    # cancellable=True: if this task is cancelled, then we abandon the
+    # thread to keep running waitpid in the background. Since this is
+    # always run as a system task, this will only happen if the whole
+    # call to trio.run is shutting down.
+
+    try:
+        await run_sync_in_worker_thread(
+            sync_wait_reapable,
+            pid,
+            cancellable=True,
+            limiter=waitid_limiter,
+        )
+    except OSError:
+        # If waitid fails, waitpid will fail too, so it still makes
+        # sense to wake up the callers of wait_process_exiting(). The
+        # most likely reason for this error in practice is a child
+        # exiting when wait() is not possible because SIGCHLD is
+        # ignored.
+        pass
+    finally:
+        event.set()
+
+
+async def wait_child_exiting(process: subprocess.Popen) -> None:
+    # Logic of this function:
+    # - The first time we get called, we create an Event and start
+    #   an instance of _waitid_system_task that will set the Event
+    #   when waitid() completes. If that Event is set before
+    #   we get cancelled, we're good.
+    # - Otherwise, a following call after the cancellation must
+    #   reuse the Event created during the first call, lest we
+    #   create an arbitrary number of threads waiting on the same
+    #   process.
+
+    ATTR = "@trio_wait_event"  # an unlikely attribute name, to be sure
```

Comment:
Currently `wait_child_exiting` takes the underlying `subprocess.Popen` object, because I didn't want to create an unnecessary circular dependency. Do you want that to change to take the `trio.Process` object?

Reply:
Ah, I missed that!

Everything here is private and maintained together – the only reason these functions aren't methods on `_subprocess.py` is to keep that file from getting too big :-). So I think we should do whatever makes the code simplest, which is probably to pass the `trio.Process` object through.

Reply:
Sounds good - I've made that change.